objects/src/builder/mod.rs

```diff
@@ -0,0 +1,266 @@
+    pub fn new(asset: AccountId, rng: T) -> Result<Self, AssetError> {
+        if !matches!(asset.account_type(), AccountType::NonFungibleFaucet) {
+            return Err(AssetError::not_a_non_fungible_faucet_id(asset));
+        }
+
+        Ok(Self { asset, rng })
+    }
+
+    pub fn build(&mut self) -> Result<NonFungibleAssetDetails, AssetError> {
+        let data = (&mut self.rng).sample_iter(Standard).take(5).collect();
+        NonFungibleAssetDetails::new(self.asset, data)
+    }
+}
+
+/// Builder for an `NonFungibleAsset`, the builder can be configured and used multipled times.
+pub struct NonFungibleAssetBuilder<T: Rng> {
+    details_builder: NonFungibleAssetDetailsBuilder<T>,
+}
+
+impl<T: Rng> NonFungibleAssetBuilder<T> {
+    pub fn new(asset: AccountId, rng: T) -> Result<Self, AssetError> {
+        let details_builder = NonFungibleAssetDetailsBuilder::new(asset, rng)?;
+        Ok(Self { details_builder })
+    }
+
+    pub fn with_details(
+        &mut self,
+        details: &NonFungibleAssetDetails,
+    ) -> Result<NonFungibleAsset, AssetError> {
+        NonFungibleAsset::new(&details)
+    }
+
+    pub fn build(&mut self) -> Result<NonFungibleAsset, AssetError> {
+        let details = self.details_builder.build()?;
+        NonFungibleAsset::new(&details)
+    }
+}
+
+/// Builder for an `Account`, the builder allows for a fluent API to construct an account. Each
+/// account needs a unique builder.
+pub struct AccountBuilder {
+    account_id: AccountId,
+    assets: Vec<Asset>,
+    storage_builder: AccountStorageBuider,
+    code: String,
+    nonce: Felt,
+}
+
+const DEFAULT_ACCOUNT_CODE: &str = "\
+use.miden::sat::account
+
+export.incr_nonce
+    push.0 swap
+    # => [value, 0]
+
+    exec.account::incr_nonce
+    # => [0]
+end
+
+export.set_item
+    exec.account::set_item
+    # => [R', V, 0, 0, 0]
+
+    movup.8 drop movup.8 drop movup.8 drop
+    # => [R', V]
+end
+
+export.set_code
+    padw swapw
+    # => [CODE_ROOT, 0, 0, 0, 0]
+
+    exec.account::set_code
+    # => [0, 0, 0, 0]
+end
+
+export.account_procedure_1
+    push.1.2
+    add
+end
+
+export.account_procedure_2
+    push.2.1
+    sub
+end
+";
+
+impl AccountBuilder {
+    pub fn new(account_id: AccountId) -> Self {
+        Self {
+            account_id,
+            assets: vec![],
+            storage_builder: AccountStorageBuider::new(),
+            code: DEFAULT_ACCOUNT_CODE.to_string(),
+            nonce: ZERO,
+        }
+    }
+
+    pub fn add_asset(mut self, asset: Asset) -> Self {
+        self.assets.push(asset);
+        self
+    }
+
+    pub fn add_storage_item(mut self, item: StorageItem) -> Self {
+        self.storage_builder.add_item(item);
+        self
+    }
+
+    pub fn code<T: AsRef<str>>(mut self, code: T) -> Self {
+        self.code = code.as_ref().to_string();
+        self
+    }
+
+    pub fn nonce(mut self, nonce: Felt) -> Self {
+        self.nonce = nonce;
+        self
+    }
```

Comment:
For the `AccountIdBuilder` the builder methods return mutable reference however here we consume the object.  Whats the reason for this discrepancy?

Reply:
makes sense thanks for the explanation.